### DISCUSSION TITLE
I get Bad state: error when calling onDispose

### DISCUSSION THREAD

**Discussion by titomitto** [rrousselGit/riverpod#2366] [Q&A] [answered]
I'm not sure what I may be doing wrong but I have this code in my app:
```dart
final tasksLocalDataSourceProvider =
    FutureProvider.autoDispose<TasksLocalDataSource>((ref) async {
  Hive.registerAdapter(TaskModelAdapter());

  Box<TaskModel> box = await Hive.openBox("tasks");

  var tasksLocalDataSource = TasksLocalDataSourceImpl(box);

  ref.onDispose(() {
    tasksLocalDataSource.close();
    log("tasksLocalDataSourceProvider cancelled");
  });

  return tasksLocalDataSource;
});

```
Is there something I am missing, coz I keep getting this exception: 
`Unhandled Exception: Bad state: Cannot call onDispose after a provider was dispose`

This is part of my open source implementation of Clean Architecture in a [todo app](https://github.com/TitoMitto/clean-todo), I've been trying to debug it the whole day. You can clone the project, if need be.

For added context here is how I consume it inside the app:
```dart
final tasksRepositoryProvider = FutureProvider<TasksRepository>((ref) async {
  final tasksLocalDataSource =
      await ref.read(tasksLocalDataSourceProvider.future);
  return TasksRepositoryImpl(localDataSource: tasksLocalDataSource);
});
```

**Comment by rrousselGit** [comment, marked as answer]
Your provider was disposed during the `await Hive.openBox("tasks");`

Make sure to watch your provider. Don't use `ref.read` like you did in `tasksRepositoryProvider`

**Reply by titomitto** [reply]
How would you have implemented it if you were using the structure I used in my app? I have tried replacing how I consume it in the `tasksRepositoryProvider` but it still throws the same error. Below is the updated code:
```dart
final tasksRepositoryProvider =
    FutureProvider.autoDispose<TasksRepository>((ref) async {
  final tasksLocalDataSource =
      await ref.watch(tasksLocalDataSourceProvider.future);
  return TasksRepositoryImpl(localDataSource: tasksLocalDataSource);
});
```

**Reply by titomitto** [reply]
Fixed it.
